File: leetcode-04/solutions/cpp/0239.cpp
class Solution {
 public:
  vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    vector<int> ans;
    deque<int> deque;

    for (int i = 0; i < nums.size(); i++) {
      while (!deque.empty() && nums[i] > deque.back()) deque.pop_back();
      deque.push_back(nums[i]);
      if (i - k + 1 >= 0) {
        ans.push_back(deque.front());
        if (nums[i - k + 1] == deque.front()) deque.pop_front();
      }
    }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0239.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []
        deque = collections.deque()

        for i in range(len(nums)):
            while deque and nums[i] > deque[-1]:
                deque.pop()
            deque.append(nums[i])
            if i - k + 1 >= 0:
                ans.append(deque[0])
                if nums[i - k + 1] == deque[0]:
                    deque.popleft()

        return ans
